ethcore/src/snapshot/consensus/mod.rs

```diff
@@ -22,3 +22,18 @@ mod work;
 
 pub use self::authority::*;
 pub use self::work::*;
+
+use ethash_engine::{MAX_SNAPSHOT_BLOCKS, SNAPSHOT_BLOCKS};
+use snapshot::SnapshotComponents;
+
+/// Create a factory for building snapshot chunks and restoring from them.
+/// `None` indicates that the engine doesn't support snapshot creation.
+pub fn chunker(engine_name: &str) -> Option<Box<dyn SnapshotComponents>> {
+	match engine_name {
+		"AuthorityRound" => Some(Box::new(PoaSnapshot)),
```

Comment:
Hmm - this feels really fragile. We do import all engines to `ethcore` anyway, right? Can't we at least use constants?

Reply:
It is, and it's the crucial point point of this PR. The thing to note is that we have two modes of performing a snapshot: PoA and PoW. Neither is tied to the specific engine used. That means there's an opportunity to decouple the two and get rid of the `snapshot_components()` method of the `Engine` trait. 
That method of the `Engine` trait is what makes it problematic to extract engines from `ethcore` because snapshots depend on `Client`.
In this PR we do import all engines to `ethcore`, but in [the next one]() `spec` is extracted. 

Using constants in a clean way is surprisingly messy. Callers of `chunker()` have a `dyn Engine` available so the obvious solution to add an associated `NAME` const to the `Engine` trait doesn't work (because trait object safety…). I did make some feeble attempts to have the `name()` method return a `TypeId` instead of a `&str` but it turned ugly fast.
Of course I could add an `ENGINES` local const here and match on that, but it didn't seem like a real improvement tbh. Happy to do that if you feel it's better.

Reply:
Had another stab at this: added an `EngineType` to `common_types` and changed `Engine.name()` to return that. Cleans up the match statement here and is neater. It still means we have to maintain a list of the engines we support but it is a bit better.

Reply:
I was rather thinking about something like this (it's not cool if we need to enumerate all engines):
```
/// authority_round.rs
const NAME: &'str = "AuthorityRound";

// snapshot/consensus/mod.rs
match name {
   authority_round::NAME => ...
}
```

However now I thought about a different approach:
```
enum Snapshotting {
  Unsupported,
  Pow { snapshot_blocks, max_snapshot_blocks },
  Poa,
}
/// + a function in Engine
Engine::supports_snapshots() -> Snapshotting;
```
So you don't even need to enumerate engines at all, and they just say what kind of snapshotting they support.

Reply:
`authority_round::NAME => ...` still means we need to manually enumerate them so that feels like more of the same.

The `Snapshotting` works though, good idea, will try that! :)